### ISSUE TITLE
🧊 [Feature] FixedUpdate Method!

### ISSUE THREAD

**Issue by MrScautHD** [MonoGame/MonoGame#7969]
Hello!
I would love if FixedUpdate come in the Framework too!
It is simple to add it by my self but it can be usefull.

Example how it would looks like:

```cs
    protected override void Update(GameTime gameTime) {
        
    }
    
/**
* You can Change the Fixed Update Ticks but by default is it 60 Ticks
*/
    protected override void FixedUpdate(GameTime gameTime) {
        
    }
    ````

**Comment by stromkos** [comment]
You could easily implement this feature within your game,  but the usefulness across all games is quite limited. Therefore, I would not expect this feature request to progress any further.

**Comment by MrScautHD** [comment]
It is easy to implement but it is a usefull feature.

You never need more then 60 Ticks
But you want more because the Drawing, so it would both possible. (It is easy to add but it would very usefull if they add it by default)

**Comment by Beyley** [comment]
> It is easy to implement but it is a usefull feature.
> 
> You never need more then 60 Ticks
> But you want more because the Drawing, so it would both possible. (It is easy to add but it would very usefull if they add it by default)

well as the previous reply stated, usefulness is quite limited in most scenarios and is at most 6 lines of code to do, so i don't expect this to be put mainline as a default feature unless there's a more general use i can't see

**Comment by MrScautHD** [comment]
> > It is easy to implement but it is a usefull feature.
> > You never need more then 60 Ticks
> > But you want more because the Drawing, so it would both possible. (It is easy to add but it would very usefull if they add it by default)
> 
> well as the previous reply stated, usefulness is quite limited in most scenarios and is at most 6 lines of code to do, so i don't expect this to be put mainline as a default feature unless there's a more general use i can't see

I mean no one like to run his game Async it would usefull.

I mean yea not everyone would use it but it would a cool small feature that would improve the Framework.

**Comment by MrScautHD** [comment]
And i mean no one hates more features? :D

**Comment by Beyley** [comment]
> And i mean no one hates more features? :D

developers having to maintain a little used feature will

**Comment by MrGrak** [comment]
actually, plenty of people do not like when something simple grows into something complex because people keep pushing their ideas into it. as has been stated twice above, this is simple to implement for your use case. arguably, this addition would increase the complexity of the framework and possibly confuse new users (since there would be multiple fixed updates to choose from).

**Comment by MrScautHD** [comment]
> actually, plenty of people do not like when something simple grows into something complex because people keep pushing their ideas into it. as has been stated twice above, this is simple to implement for your use case. arguably, this addition would increase the complexity of the framework and possibly confuse new users (since there would be multiple fixed updates to choose from).

No i not think like that.
I would love more Features and MonoGame is already very Complex (Messy).
In generel should it get a clean up and a Documentation in above the Methods. https://github.com/MonoGame/MonoGame/issues/7968

**Comment by MrScautHD** [comment]
> > And i mean no one hates more features? :D
> 
> developers having to maintain a little used feature will

I think they create it and take never again a look on it so...

**Comment by MrGrak** [comment]
you're free to fork monogame and add whatever you'd like to it. please mark this issue as closed.

**Comment by Xrayez** [comment]
@MrScautHD I think the framework is quite conservative in this regard, especially when the primary motivation behind MonoGame is to replicate XNA, so if I were you, I wouldn't expect these kind of features to be added to the framework.

**Comment by squarebananas** [comment]
I may be misunderstanding what is being requested but fixed/variable time step options are already available:
https://docs.monogame.net/api/Microsoft.Xna.Framework.Game.html#Microsoft_Xna_Framework_Game_IsFixedTimeStep

**Comment by MrScautHD** [comment]
> I may be misunderstanding what is being requested but fixed/variable time step options are already available: https://docs.monogame.net/api/Microsoft.Xna.Framework.Game.html#Microsoft_Xna_Framework_Game_IsFixedTimeStep

Yea but this allow you just always fixes updates my idea is to allow both

**Comment by squarebananas** [comment]
Fixed timestep runs a fixed number of updates and skips draws if drawing would otherwise slow down the fixed updates. If both were fixed and the combined update/draw took longer than the target time then visible slow downs would occur (as the sum of fixed elapsed times would become completely disconnected from the real time elapsed with no way to catch up).

**Comment by stromkos** [comment]
> I mean no one like to run his game Async it would usefull.
> 
> I mean yea not everyone would use it but it would a cool small feature that would improve the Framework.

It provides no benefit. Asking the community to implement a feature to support only your use case is wrong.

**Comment by MrScautHD** [comment]
> Fixed timestep runs a fixed number of updates and skips draws if drawing would otherwise slow down the fixed updates. If both were fixed and the combined update/draw took longer than the target time then visible slow downs would occur (as the sum of fixed elapsed times would become completely disconnected from the real time elapsed with no way to catch up).

Wait?

Does FixedTimeStep slow down the Draw too? or just the Update

**Comment by MrScautHD** [comment]
i think FixedTime step should then work just for the Update Method and the Draw Method get handled by Vsync

**Comment by Beyley** [comment]
> > Fixed timestep runs a fixed number of updates and skips draws if drawing would otherwise slow down the fixed updates. If both were fixed and the combined update/draw took longer than the target time then visible slow downs would occur (as the sum of fixed elapsed times would become completely disconnected from the real time elapsed with no way to catch up).
> 
> Wait?
> 
> Does FixedTimeStep slow down the Draw too? or just the Update

FixedTimeStep effects draw, as Update and Draw are intrisically tied in MonoGame, its Update -> Draw -> Update -> Draw, if you want extra things in the middle, you have to do it yourself

**Comment by stromkos** [comment]
> Does FixedTimeStep slow down the Draw too? or just the Update

Just `Update()`
The screen keeps its previous values . If the update time expires.

Fixed timing  does not slow anything down, unless the Update is running slowly. (i.e. more than 16.6 ms per frame)

If that does happen' you need to optimize your code.

**Comment by MrScautHD** [comment]
... Thats made very wierd, it should just slow down Update.

**Comment by MrScautHD** [comment]
Because allot of guys want smooth gameplay with 144 FPS...

But the tick should stay on 60 FPS to stay Synced.

**Comment by MrScautHD** [comment]
^^ For Movement, Physics....

**Comment by Beyley** [comment]
> Because allot of guys want smooth gameplay with 144 FPS...
> 
> But the tick should stay on 60 FPS to stay Synced.

Not every game needs to limit physics and movement to 60fps, infact *most* dont

**Comment by MrScautHD** [comment]
> 

But you should...

And btw if you not wanna limit it to 60 FPS just disable the FixedTimeStep.

**Comment by MrScautHD** [comment]
Small idea (To give the Users more options)

FixedUpdateTimeStep = true;
FixedDrawTimeStep = true;

**Comment by MrScautHD** [comment]
^^ you can set what witch Time Step should do

**Comment by MrScautHD** [comment]
I ReOpen this because that should really get splited.

**Comment by Beyley** [comment]
they dont need to be split for most games, and if you need a slower fixed update than draw, just have a fixed time step in your code based off the non-fixed timestep, its like at most 6 lines of code

**Comment by Beyley** [comment]
your proposal is a maintenance burden (most games dont need it, so its probably ittl acquire bitrot over time), and a breaking change for the sake of a breaking change

**Comment by squarebananas** [comment]
> Wait?
> 
> Does FixedTimeStep slow down the Draw too? or just the Update

Yes, FixedTimeStep can skip Draws. FixedTimeStep will attempt to update at a fixed rate while the time elapsed between draws will be greater after draw skips.

I would recommend reading this blog post by Shawn Hargreaves (one of the original XNA developers) which explains the differences between fixed and variable timesteps and the reasoning behind the way they are implemented. Actually a lot of his blog posts are great for understanding the way in which XNA was designed and many principles are still relevant for MonoGame.
https://shawnhargreaves.com/blog/understanding-gametime.html
